pkg/helm/chartproxy/repos.go

```diff
@@ -0,0 +1,224 @@
+}
+
+type HelmRepoGetter interface {
+	List() ([]*helmRepo, error)
+}
+
+type helmRepoGetter struct {
+	Client     dynamic.Interface
+	CoreClient corev1.CoreV1Interface
+}
+
+func (b helmRepoGetter) unmarshallConfig(repo unstructured.Unstructured) (*helmRepo, error) {
+	h := &helmRepo{}
+	urlValue, _, err := unstructured.NestedString(repo.Object, "spec", "connectionConfig", "url")
+	if err != nil {
+		return nil, err
+	}
+
+	h.Url, err = url.Parse(urlValue)
+	if err != nil {
+		return nil, err
+	}
+
+	h.Name, _, err = unstructured.NestedString(repo.Object, "metadata", "name")
+	if err != nil {
+		return nil, err
+	}
+
+	caReference, _, err := unstructured.NestedString(repo.Object, "spec", "connectionConfig", "ca", "name")
+	if err != nil {
+		return nil, err
+	}
+
+	tlsReference, _, err := unstructured.NestedString(repo.Object, "spec", "connectionConfig", "tlsClientConfig", "name")
+	if err != nil {
+		return nil, err
+	}
+
+	var rootCAs *x509.CertPool
+	if caReference != "" {
+		caCert, err := b.configMapValue(caReference, "ca-bundle.crt")
+		if err != nil {
+			return nil, err
+		}
+		if caCert != nil {
+			rootCAs = x509.NewCertPool()
+			if ok := rootCAs.AppendCertsFromPEM(caCert); !ok {
+				return nil, errors.New("Failed to append caCert")
+			}
+		}
+	}
+	if rootCAs == nil {
+		rootCAs, err = x509.SystemCertPool()
+		if err != nil {
+			return nil, err
+		}
+	}
+	h.TlsClientConfig = &tls.Config{
+		RootCAs:      rootCAs,
+		CipherSuites: crypto.DefaultCiphers(),
+	}
+	if tlsReference != "" {
+		tlsCert, err := b.secretValue(tlsReference, "tls.crt")
+		if err != nil {
+			return nil, err
+		}
+		tlsKey, err := b.secretValue(tlsReference, "tls.key")
+		if err != nil {
+			return nil, err
+		}
+		if tlsKey != nil && tlsCert != nil {
+			cert, err := tls.X509KeyPair(tlsCert, tlsKey)
+			if err != nil {
+				return nil, err
+			}
+			h.TlsClientConfig.Certificates = []tls.Certificate{cert}
+		}
+	}
+	return h, nil
+}
+
+func mergeIndexFiles(files ...*repo.IndexFile) *repo.IndexFile {
+	indexFile := repo.NewIndexFile()
+	for _, file := range files {
+		for key, entry := range file.Entries {
+			indexFile.Entries[key] = entry
+		}
+	}
+	return indexFile
+}
+
+func (b helmRepoGetter) configMapValue(name string, dataField string) ([]byte, error) {
+	configMap, err := b.CoreClient.ConfigMaps(configNamespace).Get(context.TODO(), name, v1.GetOptions{})
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Failed to find %s key in configmap %s", dataField, name))
+	}
+	if val, ok := configMap.Data[dataField]; ok {
+		return []byte(val), nil
+	}
+	return nil, errors.New(fmt.Sprintf("Failed to find %s key in configmap %s", dataField, name))
+}
+
+func (b helmRepoGetter) secretValue(name string, dataField string) ([]byte, error) {
+	secret, err := b.CoreClient.Secrets(configNamespace).Get(context.TODO(), name, v1.GetOptions{})
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Failed to find %s key in secret %s", dataField, name))
```

Comment:
This error is not due to the missing key in the secret but rather an API call issue